pkg/dataobj/sections/internal/columnar/encoder.go

```diff
@@ -0,0 +1,345 @@
+		Compression:      desc.Compression,
+		UncompressedSize: uint64(desc.UncompressedSize),
+		CompressedSize:   uint64(desc.CompressedSize),
+		Statistics:       desc.Statistics,
+
+		// These fields aren't available until the column is closed. We
+		// temporarily set them to the maximum values so they're accounted for
+		// when estimating metadata size.
+
+		PagesCount:           math.MaxUint32,
+		ColumnMetadataOffset: math.MaxUint32,
+		ColumnMetadataLength: math.MaxUint32,
+	}
+
+	return newColumnEncoder(enc, enc.dataSize()), nil
+}
+
+func (enc *Encoder) getDictionaryRef(text string) uint32 {
+	enc.initDictionary() // Make sure the dictionary is initialized.
+
+	if idx, ok := enc.dictionaryLookup[text]; ok {
+		return idx
+	}
+
+	idx := uint32(len(enc.dictionary))
+	enc.dictionary = append(enc.dictionary, text)
+	enc.dictionaryLookup[text] = idx
+
+	return idx
+}
+
+func (enc *Encoder) initDictionary() {
+	if len(enc.dictionary) > 0 && len(enc.dictionaryLookup) > 0 {
+		return // Already initialized.
+	}
+
+	// Initialize the dictionary with index zero being reserved for the
+	// empty string.
+	enc.dictionary = []string{""}
+	enc.dictionaryLookup = map[string]uint32{"": 0}
+}
+
+// dataSize returns the current number of buffered data bytes.
+func (enc *Encoder) dataSize() int {
+	if enc.data == nil {
+		return 0
+	}
+	return enc.data.Len()
+}
+
+// append adds column data and metadata to enc. append must only be called from
+// child elements on Close or Discard. Discard calls must pass nil for both data
+// and metadata to denote a discard.
+//
+// enc *must never* retain data or metadata beyond the call to append.
+func (enc *Encoder) append(pages int, data, metadata []byte) error {
+	if enc.curColumn == nil {
+		return errElementNoExist
+	}
+
+	if len(data) == 0 && len(metadata) == 0 {
+		// Column was discarded.
+		enc.curColumn = nil
+		return nil
+	}
+
+	// Initialize enc.data and enc.metadata.
+	enc.initBuffers()
+
+	// Update deferred fields now that we know their values.
+	enc.curColumn.PagesCount = uint64(pages)
+	enc.curColumn.ColumnMetadataOffset = uint64(enc.metadata.Len())
+	enc.curColumn.ColumnMetadataLength = uint64(len(metadata))
+
+	// [bytes.Buffer.Write] never fails.
+	_, _ = enc.data.Write(data)
+	_, _ = enc.metadata.Write(metadata)
+
+	enc.columns = append(enc.columns, enc.curColumn)
+	enc.curColumn = nil
+	return nil
+}
+
+func (enc *Encoder) initBuffers() {
+	if enc.data != nil && enc.metadata != nil {
+		return
+	}
+
+	enc.data = bufpool.GetUnsized()
+	enc.metadata = bufpool.GetUnsized()
+
+	// Initialize the metadata buffer with the format version. Section implementations
+	// will use the version stored in the SectionType, but we write the format
+	// version to the metadata anyway so that older versions of dataobj code
+	// (where the version was encoded in metadata) recognizes that it's a newer format.
+	//
+	// TODO(rfratto): remove this check once we've fully rolled out the new
```

Comment:
Can this be removed completely if its a breaking change?

Reply:
It can be, though with some minor risk as we roll this out: if the first byte is `0x01`, then old queriers that are still running as we roll out the new weekly will think that the newer objects are supported, and they may crash since they'll start reading invalid data. 

I figured this was a low cost guard against that: it's a single byte per section that we never read, and it can be removed without a breaking change or a new version (since it's never read back).